File: android_client/android/app/src/main/java/ru/smarthome/ControllerViewHolder.java
package ru.smarthome;

import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.View;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import ru.smarthome.library.ArduinoDevice;
import ru.smarthome.library.Controller;
import ru.smarthome.library.RaspberryResponse;

import static ru.smarthome.MainActivity.getRaspberryApi;

public class ControllerViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
    public static final String TAG = ControllerViewHolder.class.getSimpleName();
    public static final String UNKNOWN_STATE = "-";
    private TextView guid;
    private TextView type;
    private TextView state;
    private ProgressBar progressBar;

    private ArduinoDevice device;
    private Controller controller;

    public ControllerViewHolder(View itemView) {
        super(itemView);

        guid = itemView.findViewById(R.id.controller_guid);
        type = itemView.findViewById(R.id.type);
        state = itemView.findViewById(R.id.state);
        progressBar = itemView.findViewById(R.id.progress);

        itemView.setOnClickListener(this);
    }

    public void bind(ArduinoDevice device, Controller controller) {
        if (controller == null || device == null) {
            return;
        }
        this.controller = controller;
        this.device = device;

        guid.setText(controller.guid + "");
        type.setText(controller.type);
        if (controller.state != null) {
            state.setText(controller.state);
        } else {
            state.setText(UNKNOWN_STATE);
        }
    }

    @Override
    public void onClick(View v) {
        if (controller.type.equals("ARDUINO_ON_OFF")) {
            if (UNKNOWN_STATE.equals(state.getText())) {
                readState();
                return;
            }
            if ("1".equals(controller.state)) {
                changeStateTo("0");
            } else {
                changeStateTo("1");
            }

        } else {
            readState();
        }
    }

    private void startStateChange() {
        progressBar.setVisibility(View.VISIBLE);
        state.setVisibility(View.GONE);
    }

    private void endStateChange() {
        progressBar.setVisibility(View.GONE);
        state.setVisibility(View.VISIBLE);
    }

    private void readState() {
        startStateChange();
        RaspberryApi api = getRaspberryApi();
        Call<RaspberryResponse> call = api.readControllerState(device.guid, controller.guid);
        call.enqueue(new Callback<RaspberryResponse>() {
            @Override
            public void onResponse(Call<RaspberryResponse> call, Response<RaspberryResponse> response) {
                handleResponseWithState(response);
            }

            @Override
            public void onFailure(Call<RaspberryResponse> call, Throwable t) {
                Log.d(TAG, "onFailure: " + t);
                endStateChange();
            }
        });

    }

    private void changeStateTo(String value) {
        startStateChange();
        RaspberryApi api = getRaspberryApi();
        Call<RaspberryResponse> call = api.changeControllerState(device.guid, controller.guid, value);
        call.enqueue(new Callback<RaspberryResponse>() {
            @Override
            public void onResponse(Call<RaspberryResponse> call, Response<RaspberryResponse> response) {
                handleResponseWithState(response);
            }

            @Override
            public void onFailure(Call<RaspberryResponse> call, Throwable t) {
                Log.d(TAG, "onFailure: " + t);
                endStateChange();
            }
        });

    }

    private void handleResponseWithState(Response<RaspberryResponse> response) {
        if (response.isSuccessful()) {

            RaspberryResponse raspberryResponse = response.body();
            if (raspberryResponse == null) {
                endStateChange();
                return;
            }
            String newState = raspberryResponse.response;
            state.setText(newState);
            controller.state = newState;
        } else {
            String message = "Returned code: " + response.code();
            try {
                message += ", " + response.raw().body().string();
            } catch (Exception ignored) {}
            
            Toast.makeText(guid.getContext(), message, Toast.LENGTH_LONG).show();
        }
        endStateChange();
    }
}

File: android_client/android/app/src/main/java/ru/smarthome/ControllerViewHolder.kt
package ru.smarthome

import android.support.v7.widget.RecyclerView
import android.util.Log
import android.view.View
import android.widget.ProgressBar
import android.widget.TextView
import android.widget.Toast

import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import ru.smarthome.MainActivity.Companion.raspberryApi
import ru.smarthome.library.BaseController
import ru.smarthome.library.ControllerType
import ru.smarthome.library.IotDevice
import ru.smarthome.library.RaspberryResponse

class ControllerViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView), View.OnClickListener {
    val TAG = "ControllerViewHolder"
    val UNKNOWN_STATE = "-"

    private val guid: TextView = itemView.findViewById(R.id.controller_guid)
    private val type: TextView = itemView.findViewById(R.id.type)
    private val state: TextView = itemView.findViewById(R.id.state)
    private val progressBar: ProgressBar = itemView.findViewById(R.id.progress)

    private var device: IotDevice? = null
    private var controller: BaseController? = null

    init {
        itemView.setOnClickListener(this)
    }

    fun bind(device: IotDevice?, controller: BaseController?) {
        this.controller = controller
        this.device = device

        guid.text = "${controller?.guid}"
        type.text = "${controller?.type}"
        state.text = controller?.state ?: UNKNOWN_STATE
    }

    override fun onClick(v: View) {
        if (controller?.type == ControllerType.ARDUINO_ON_OFF) {
            if (state.text.toString() == UNKNOWN_STATE) {
                readState()
                return
            }

            when (controller?.state) {
                "0" -> changeStateTo("1")
                "1" -> changeStateTo("0")
            }
        } else {
            readState()
        }
    }

    private fun startStateChange() {
        progressBar.visibility = View.VISIBLE
        state.visibility = View.GONE
    }

    private fun endStateChange() {
        progressBar.visibility = View.GONE
        state.visibility = View.VISIBLE
    }

    private fun readState() {
        startStateChange()

        val deviceGuid = device?.guid ?: return
        val controllerGuid = controller?.guid ?: return

        raspberryApi.readControllerState(deviceGuid, controllerGuid).enqueue {
            onResponse = {
                handleResponseWithState(it)
            }

            onFailure = {
                Log.d(TAG, "onFailure: $it")
                endStateChange()
            }
        }
    }

    private fun changeStateTo(value: String) {
        startStateChange()

        val deviceGuid = device?.guid ?: return
        val controllerGuid = controller?.guid ?: return

        raspberryApi.changeControllerState(deviceGuid, controllerGuid, value).enqueue {
            onResponse = {
                handleResponseWithState(it)
            }

            onFailure = {
                Log.d(TAG, "onFailure: $it")
                endStateChange()
            }
        }
    }

    private fun handleResponseWithState(response: Response<RaspberryResponse>) {
        if (response.isSuccessful) {
            response.body()?.let {
                state.text = it.response
                controller?.state = it.response
            }
        } else {
            val message = "Returned code: ${response.code()}, body=${response.raw().body()?.string()}"
            Toast.makeText(guid.context, message, Toast.LENGTH_LONG).show()
        }

        endStateChange()
    }
}

